### ISSUE TITLE
TestGenerateBlockAndImportEthash should pass

### ISSUE THREAD

**Issue by lightsing** [scroll-tech/go-ethereum#53]
https://github.com/scroll-tech/go-ethereum/blob/97faa8cc63d704a11c29d2384e1d450ceec4d888/miner/worker_test.go#L207-L210

currently failed with log:
```
=== RUN   TestGenerateBlockAndImportEthash

    worker_test.go:265: failed to insert new mined block 1: invalid receipt root hash (remote: 46be8924226c54f771d30283b863add0b365f39298253ce44601c035db6323ee local: db85ddb0e660e1c591341cb76012b4ae5a384b62322e196e387bbf3ba8dcab59)
--- FAIL: TestGenerateBlockAndImportEthash (0.01s)

FAIL
```

Priority

P0 TestGenerateBlockAndImportEthash
P0 TestTransactionIndices

**Comment by lispc** [comment]
and also TestTransactionIndices

**Comment by hsyodyssey** [comment]
I have an idea about `TestGenerateBlockAndImportEthash`. This problem may be caused by the inconsistency of the poststate in the recepit generated by the Block generation phase and the recepit generated by the Block verification phase.

https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/state_processor.go#L108-L112

```
[I!] Receipt PostState in Worker:  0x2afe449e3e47f8a6d962a5836f7037957a927f02350b179e23360bd5e3e98022
[I!] Receipt PostState in Worker:  0x1528d204ed92ea53edd51d3fbca7a4d52f0c55757c1f57999a6fe0ed786c3bcd

[I!] Receipt PostState in Process:  0x0a8b311944285b92b9d8a303f8b61b375dc6aad9e844580be3d2f7cd7fa2fddb
[I!] Receipt PostState in Process:  0x035a527a16856cd0a093741bae166a6912d97aefea518fc0166551681d609c92
```

**Comment by hsyodyssey** [comment]
For TestGenerateBlockAndImportEthash, the test also fails here: https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/block_validator.go#L99.

**Comment by hsyodyssey** [comment]
For TestTransactionIndices, I think there are some anomalies that occur in https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain.go#L1471-L1472.

lastCanon should not be nil here. Since the value of lastCanon is nil, the data of the Block is not transferred to the Channel, causing the subsequent exceptions. looking at what is causing this phenomenon.

**Comment by hsyodyssey** [comment]
Oh, I found the problem. Since the logical call relationship of the functions here is rather long, I will only mention the key calles here.

The problem occurs when the test case calls insertChain for the first time (In this test: Block number 129).

1. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain.go#L1488
`block, err := it.next()`

2. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain_insert.go#L121
`return it.chain[it.index], it.validator.ValidateBody(it.chain[it.index])
`
3. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/block_validator.go#L68-L73
In the test case, when the first block insertion is made, an error of `consensus.ErrPrunedAncestor` should be returned to trigger the subsequent logic. But the current code returns nil here.
```
	if !v.bc.HasBlockAndState(block.ParentHash(), block.NumberU64()-1) {
		if !v.bc.HasBlock(block.ParentHash(), block.NumberU64()-1) {
			return consensus.ErrUnknownAncestor
		}
		return consensus.ErrPrunedAncestor
	}
```
4. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain_reader.go#L262-L265
The deeper reason for not returning `consensus.ErrPrunedAncestor` is that the call to `bc.stateCache.OpenTrie` in `HasState` returns a nil value. If we compare with the original Geth, we will see that they return a `trie.MissingNodeError` when calling` bc.stateCache.OpenTrie`.

```
func (bc *BlockChain) HasState(hash common.Hash) bool {
	_, err := bc.stateCache.OpenTrie(hash)
	return err == nil
}
```

In summary, these minor deviations eventually lead to incorrect logic execution, such as the above mentioned value of `lastCanon` becoming `nil`. 

https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/trie/secure_binary_trie.go#L51-L63

Exception handling in the place of creating a new SecureBinaryTrie might also need some considerations?

**Comment by lispc** [comment]
> Oh, I found the problem. Since the logical call relationship of the functions here is rather long, I will only mention the key calles here.
> 
> The problem occurs when the test case calls insertChain for the first time (In this test: Block number 129).
> 
> 1. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain.go#L1488
>    `block, err := it.next()`
> 2. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain_insert.go#L121
>    `return it.chain[it.index], it.validator.ValidateBody(it.chain[it.index]) `
> 3. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/block_validator.go#L68-L73
>    In the test case, when the first block insertion is made, an error of `consensus.ErrPrunedAncestor` should be returned to trigger the subsequent logic. But the current code returns nil here.
> 
> ```
> 	if !v.bc.HasBlockAndState(block.ParentHash(), block.NumberU64()-1) {
> 		if !v.bc.HasBlock(block.ParentHash(), block.NumberU64()-1) {
> 			return consensus.ErrUnknownAncestor
> 		}
> 		return consensus.ErrPrunedAncestor
> 	}
> ```
> 
> 4. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/blockchain_reader.go#L262-L265
>    The deeper reason for not returning `consensus.ErrPrunedAncestor` is that the call to `bc.stateCache.OpenTrie` in `HasState` returns a nil value. If we compare with the original Geth, we will see that they return a `trie.MissingNodeError` when calling` bc.stateCache.OpenTrie`.
> 
> ```
> func (bc *BlockChain) HasState(hash common.Hash) bool {
> 	_, err := bc.stateCache.OpenTrie(hash)
> 	return err == nil
> }
> ```
> 
> In summary, these minor deviations eventually lead to incorrect logic execution, such as the above mentioned value of `lastCanon` becoming `nil`.
> 
> https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/trie/secure_binary_trie.go#L51-L63
> 
> Exception handling in the place of creating a new SecureBinaryTrie might also need some considerations?

Great analysis! Fixed now.
